File: packages/sdk-py/foxhound/transport/compression.py
"""Wire-format-agnostic compression (WP05).

Mirrors ``packages/sdk/src/transport/compression.ts``. Gzip is the
default algorithm and uses the Python stdlib ``gzip`` module (no
additional dep). LZ4 is reserved in the type signature for the
low-CPU case (see RFC-005); the implementation is guarded on the
optional ``lz4`` PyPI dep, which operators install explicitly when
they want the low-CPU wire path.

Design rules (parity with the TS side):

1. Compression is all-or-nothing per batch. We do not split batches
   across compressed chunks; the SDK already batches before transport.

2. Tiny bodies skip compression. Below ``COMPRESSION_THRESHOLD_BYTES``
   (512 B uncompressed) gzip is net-negative; ``compress()`` returns
   the input unchanged and reports ``kind == "none"``.

3. ``compress()`` may downgrade the algorithm on fallback (e.g. LZ4
   requested without the dep). The caller MUST read the returned
   ``kind`` when setting the outgoing ``Content-Encoding`` header.
"""

from __future__ import annotations

import gzip
import warnings
from dataclasses import dataclass
from typing import Literal

CompressionKind = Literal["gzip", "lz4", "none"]

COMPRESSION_THRESHOLD_BYTES = 512
"""Below this many bytes, gzip header overhead beats compression savings."""


@dataclass(frozen=True)
class CompressedBody:
    """Compressed output paired with the algorithm that produced it."""

    bytes_: bytes
    kind: CompressionKind


_lz4_warning_emitted = False


def compress(body: bytes, kind: CompressionKind) -> CompressedBody:
    """Compress ``body`` with ``kind``. Returns the body unchanged when
    the algorithm is ``none``, when the body is below the compression
    threshold, or when ``lz4`` was requested without the optional dep.

    The caller MUST read ``result.kind`` (not the input ``kind``
    argument) to set the outgoing ``Content-Encoding`` header: the
    fallback paths may have downgraded the algorithm.
    """
    if kind == "none" or len(body) < COMPRESSION_THRESHOLD_BYTES:
        return CompressedBody(bytes_=body, kind="none")
    if kind == "gzip":
        # compresslevel=6 is Python's default and matches Node's zlib
        # default; parity matters for deterministic ratio tests.
        return CompressedBody(bytes_=gzip.compress(body, compresslevel=6), kind="gzip")
    # kind == "lz4". Import lazily so the dep is optional.
    try:  # pragma: no cover - tested via fallback branch
        import lz4.frame  # type: ignore[import-not-found]

        return CompressedBody(bytes_=lz4.frame.compress(body), kind="lz4")
    except ImportError:
        global _lz4_warning_emitted
        if not _lz4_warning_emitted:
            _lz4_warning_emitted = True
            warnings.warn(
                "foxhound.transport.compression: lz4 requested but the "
                "optional 'lz4' package is not installed; falling back to "
                "uncompressed. Install the `lz4` package or switch to gzip.",
                RuntimeWarning,
                stacklevel=2,
            )
        return CompressedBody(bytes_=body, kind="none")


def decompress(body: bytes, kind: CompressionKind) -> bytes:
    """Decompress ``body`` using the algorithm signaled by ``kind``.

    ``none`` is a passthrough. Invalid gzip raises
    :class:`gzip.BadGzipFile`. ``lz4`` without the optional dep also
    falls back to passthrough with a one-time warning; callers that
    need a hard error should check the source-side encoding explicitly
    before invoking decompress.
    """
    if kind == "none":
        return body
    if kind == "gzip":
        return gzip.decompress(body)
    try:  # pragma: no cover
        import lz4.frame  # type: ignore[import-not-found]

        return lz4.frame.decompress(body)
    except ImportError:
        global _lz4_warning_emitted
        if not _lz4_warning_emitted:
            _lz4_warning_emitted = True
            warnings.warn(
                "foxhound.transport.compression: lz4 decompression "
                "requested but the optional 'lz4' package is not "
                "installed; returning body unchanged.",
                RuntimeWarning,
                stacklevel=2,
            )
        return body


def compression_kind_from_header(header: str | None) -> CompressionKind:
    """Map a ``Content-Encoding`` header value to a ``CompressionKind``.

    Unknown values (``deflate``, ``br``, missing) map to ``none``.
    """
    if not header:
        return "none"
    v = header.strip().lower()
    if v == "gzip":
        return "gzip"
    if v == "lz4":
        return "lz4"
    return "none"


def _reset_lz4_warning_for_tests() -> None:
    """Expose a hook so tests can re-arm the one-time warning."""
    global _lz4_warning_emitted
    _lz4_warning_emitted = False

File: packages/sdk/src/transport/compression.ts
/**
 * Wire-format-agnostic compression (WP05).
 *
 * The Foxhound SDK compresses its outbound batch body before egress. Gzip
 * is the default; LZ4 is reserved in the type signature for the low-CPU
 * case (see RFC-005) but the implementation is gated on an optional
 * dep (`lz4-napi`) that ships separately. Until that dep is installed,
 * LZ4 falls back to `none` and logs a one-time warning, preserving the
 * caller-visible API contract.
 *
 * Ground rules this file enforces:
 *
 * 1. Compression is all-or-nothing per batch. We do not split a batch
 *    across compressed chunks; the SDK already batches before calling
 *    transport, so a batch is the natural compression unit.
 *
 * 2. Tiny bodies skip compression. Below `COMPRESSION_THRESHOLD_BYTES`
 *    (512 B uncompressed), gzip is net-negative on wire bytes because
 *    of the gzip header overhead (~18 B plus dictionary). The
 *    `compress()` helper returns the input unchanged for these cases
 *    and signals it via the returned `{ kind: "none" }`.
 *
 * 3. Decompression is tolerant. A caller on the receiving end inspects
 *    the `Content-Encoding` header; this module exposes `decompress()`
 *    that takes the header value and does the right thing.
 *
 * 4. No allocation in the happy path. Node's `zlib.gzipSync` returns a
 *    `Buffer` directly over the compressed bytes; we wrap it in a
 *    Uint8Array view rather than copy. Same for `gunzipSync`.
 */
import { gzipSync, gunzipSync } from "node:zlib";

/**
 * Compression algorithms the SDK knows about.
 *
 * - `gzip` — Node stdlib, zero extra deps, default.
 * - `lz4`  — Reserved. Requires `lz4-napi` (optional peer). Falls back
 *            to `none` with a one-time warning when the dep is missing.
 * - `none` — Raw bytes. Always supported.
 */
export type CompressionKind = "gzip" | "lz4" | "none";

/**
 * Compressed output, paired with the algorithm that produced it. The
 * caller uses `kind` to set `Content-Encoding` on the wire.
 */
export interface CompressedBody {
  readonly bytes: Uint8Array;
  readonly kind: CompressionKind;
}

/**
 * Below this many uncompressed bytes, gzip costs more header overhead
 * than it saves; we short-circuit to `none`. The number is tuned against
 * a realistic 256-byte heartbeat batch (~8 % larger after gzip).
 */
export const COMPRESSION_THRESHOLD_BYTES = 512;

let lz4WarningEmitted = false;

/**
 * Compress `body` with the requested algorithm. Returns the body
 * unchanged when the algorithm is `none`, when the body is below the
 * compression threshold, or when LZ4 was requested but its optional
 * dep is not installed.
 *
 * The caller MUST read `result.kind` (not the input `kind` argument) to
 * decide the outgoing `Content-Encoding` header, because the fallback
 * paths may have downgraded the algorithm.
 */
export function compress(body: Uint8Array, kind: CompressionKind): CompressedBody {
  if (kind === "none" || body.byteLength < COMPRESSION_THRESHOLD_BYTES) {
    return { bytes: body, kind: "none" };
  }
  if (kind === "gzip") {
    // `gzipSync` returns a Node Buffer; Buffer is a subclass of
    // Uint8Array at runtime, so the cast is free and the downstream
    // `fetch` body accepts either.
    const out = gzipSync(body);
    return { bytes: out, kind: "gzip" };
  }
  // kind === "lz4" — guarded. `require`ing an optional dep from ESM is
  // awkward across bundlers; the clean path is to keep the hook
  // deliberate and let operators install `lz4-napi` explicitly when
  // they want the low-CPU path (RFC-005 "Compression choice" section).
  // Until then, fall back to raw bytes and warn once.
  if (!lz4WarningEmitted) {
    lz4WarningEmitted = true;
    console.warn(
      "[foxhound/compression] LZ4 requested but lz4-napi is not installed; " +
        "falling back to uncompressed. Install lz4-napi or switch to 'gzip'.",
    );
  }
  return { bytes: body, kind: "none" };
}

/**
 * Decompress `body` using the algorithm signaled by `kind`. `none` is a
 * passthrough. `lz4` falls back to passthrough when the optional dep is
 * missing (and the caller is responsible for erroring if that's not
 * acceptable). Invalid gzip raises.
 */
export function decompress(body: Uint8Array, kind: CompressionKind): Uint8Array {
  if (kind === "none") return body;
  if (kind === "gzip") {
    const out = gunzipSync(body);
    return out;
  }
  // kind === "lz4": same pragmatic fallback as above.
  if (!lz4WarningEmitted) {
    lz4WarningEmitted = true;
    console.warn(
      "[foxhound/compression] LZ4 decompression requested but lz4-napi is " +
        "not installed; returning body unchanged. This is probably a bug — " +
        "install lz4-napi or have the sender use 'gzip'.",
    );
  }
  return body;
}

/**
 * Map a `Content-Encoding` header value to a `CompressionKind`. Only
 * the algorithms the SDK knows about are recognised; anything else
 * (e.g. `deflate`, `br`) is `none`. The server side uses this to
 * negotiate what to decompress.
 */
export function compressionKindFromHeader(header: string | undefined): CompressionKind {
  if (!header) return "none";
  const v = header.toLowerCase().trim();
  if (v === "gzip") return "gzip";
  if (v === "lz4") return "lz4";
  return "none";
}

/** Expose a hook so tests can re-arm the one-time LZ4 warning. */
export function _resetLz4WarningForTests(): void {
  lz4WarningEmitted = false;
}
